README.md
# Cookie Island
A 2D game based off Terraria and Minecraft. It is a 2D sandbox game similar
to Terreria. This game has two gameplay mode: survival and a special mode.

### Project Bulletin Board
- Add gravity and jumping

### Requirements
- If you can run .exe file

### How to run
- Run the Cookie Island.exe file

### How to play
- Press WASD to move

### Function Versions
- Block Data System v3.0
- Terrain Generator v3.0
- Collision System v1.0
- Break/Place System v1.0


CHANGELOG.md
## Indev 1.3.0 - Dec 15, 2018
-[ADDED]
  - [GAME] Now you can remove blocks!

## Indev 1.2.0a - Dec 7, 2018
- [CHANGE]
  - [CODE] Clean codes and reformat the block data.
  - [CODE] Revamp loadTexture();

## Indev 1.2.0 - Dec 5, 2018
- [ADDITION]
  - [TECHINCAL] Added collision system v1.0
  - [TECHINCAL] More data value to each block assigned to be more organized
  - [TEXTURES] Added more textures (not used)
  - [CODE] Moved all player position/size constant under Game Config

## Indev 1.1.2a - Dec 5, 2018
- [NOTE] I decided to not to publicize my source code! And the license will be mere useful!

## Indev 1.1.2 - Dec 1, 2018
- Bug fixes!

## Indev 1.1.1 - Dec 1, 2018
- [FIX]
  - [TERRAIN] The terrain crashes/broke at specific point when either coordinate exceeds -80
- [NOTE] While the bug was found and fixed, I was building the collision. So anything you saw unusual is in progress.

## Indev 1.1.0 - Dec 1, 2018
- [ADDITION]
  - [PLAYER] Added player starting position

## Indev 1.0.1 - Nov 30, 2018
- [FIX] [GRAPHIC] The display problem (at ofs: 10, 10) crashes fixed :)

## Indev 1.0a - Nov 27, 2018
- Update Text File

## Indev 1.0 - Nov 27, 2018
- Initial Commit
- NOTE: Project already started


Cookie Island/Player.h
#ifndef HHH_PLAYER_H
#define HHH_PLAYER_H

#include <iostream>
#include <SFML/Graphics.hpp>

#include "Constants.h"

using namespace std;
using namespace sf;

class Player {
private:
    RectangleShape entity;
    Vector2f pos;
    Vector2f mot;
    Texture txtr;
public:
    Player();  // texture

    inline void show(RenderWindow& window);
    inline void update();
    inline void collision(RenderWindow& w, Vector2f _pos);

    void setPos(Vector2f& p);

    Vector2f getPos();
    Vector2f getMot();
};

Player::Player() {
  entity = RectangleShape(PLAYER_SIZE);
  entity.setPosition(PLAYER_ABS_POS); // This is the absolute position for the game (in center).
  pos = Vector2f(0, 0);  // This is the position of the player (but technically it is the terrain's position)

  if (!txtr.loadFromFile(RES_DIR+"no-texture.png")) {   // Loads the texture
      if (!txtr.loadFromFile(RES_DIR+"no-texture.png")) {
          cout << "[ERROR] Failed to load texture!\n";
          entity.setFillColor(Color(255, 0, 255));
      }
  } else {
      cout << "Textured loaded successfully\n";
      entity.setTexture(&txtr);
  }
}

void Player::show(RenderWindow& window) {
  window.draw(entity);
}

void Player::update() {

  if (Keyboard::isKeyPressed(Keyboard::W)) {  // get the keyboard input for movement
    pos.y += SPEED;
    mot.y = 1;
  } else if (Keyboard::isKeyPressed(Keyboard::S)) {
    pos.y -= SPEED;
    mot.y = -1;
  } else {
    mot.y = 0;
  }

  if (Keyboard::isKeyPressed(Keyboard::A)) {
    pos.x += SPEED;
    mot.x = 1;
  } else if (Keyboard::isKeyPressed(Keyboard::D)) {
    pos.x -= SPEED;
    mot.x = -1;
  } else {
    mot.x = 0;
  }

}

void Player::collision(RenderWindow& w, Vector2f _pos) {
  int LN_SZ = 5;
  int PBX = PLAYER_COLLISION_BOX.x*BLOCK_SZ;
  int PBY = PLAYER_COLLISION_BOX.y*BLOCK_SZ;
  Vector2f ADD(round(PLAYER_ABS_POS.x/BLOCK_SZ)*BLOCK_SZ-round(PLAYER_COLLISION_BOX.x/2), round(PLAYER_ABS_POS.y/BLOCK_SZ)*BLOCK_SZ-round(PLAYER_COLLISION_BOX.y/2));
  RectangleShape rline(Vector2f(LN_SZ, PBY));
  RectangleShape lline(Vector2f(LN_SZ, PBY));
  RectangleShape tline(Vector2f(PBX, LN_SZ));
  RectangleShape bline(Vector2f(PBX, LN_SZ));
  rline.setPosition(Vector2f(PBX+_pos.x+ADD.x, 0+_pos.y+ADD.y));
  lline.setPosition(Vector2f(0+_pos.x+ADD.x, 0+_pos.y+ADD.y));
  tline.setPosition(Vector2f(0+_pos.x+ADD.x, 0+_pos.y+ADD.y));
  bline.setPosition(Vector2f(0+_pos.x+ADD.x, PBY+_pos.y+ADD.y));
  w.draw(rline);
  w.draw(lline);
  w.draw(tline);
  w.draw(bline);
}

void Player::setPos(Vector2f& p) {
    pos = p;
}

Vector2f Player::getPos() {
    return pos;
}

Vector2f Player::getMot() {
    return mot;
}



#endif // PLAYER_H


Cookie Island/main.cpp
#include <iostream>
#include <SFML/Graphics.hpp>

#include "Constants.h"
#include "Terrain.h"
#include "Player.h"

using namespace std;
using namespace sf;


int main()
{
    cout << "Initializing...\n";
    //err().rdbuf(NULL);  // No error message (especially with the error message file can't load)
    Vector2i WORLD_SIZE(100, 50);  // x, y
    Vector2i PLAYER_START_POS(-20, -20);  // Player starting location (-ofs, -ofs) as where in Quad IV
    Vector2f ply_movement;

    Terrain terrain(WORLD_SIZE, PLAYER_START_POS);
    Player player;

    terrain.loadTexture();
    terrain.generateTerrain();
    window.setFramerateLimit(FPS_LIMIT);

    cout << "Game Started!\n";

    while (window.isOpen())
    {
        Event event;
        while (window.pollEvent(event))
        {
            if (event.type == Event::Closed)
                window.close();
        }
        ply_movement = player.getMot();

        window.clear();
        player.update();

        terrain.setPos(ply_movement);
        terrain.showTerrain();

        player.show(window);
        player.collision(window, terrain.getPos());
        window.display();

    }
    cout << "Game Ended!\n";
    return 0;
}


Cookie Island/Terrain.h
#ifndef HHH_TERRAIN_H
#define HHH_TERRAIN_H

#include <iostream>
#include <vector>
#include <SFML/Graphics.hpp>

#include "Constants.h"
#include "BlockData.h"

using namespace std;
using namespace sf;

class Terrain {
private:
    Vector2i TERRAIN_SZ;
    Vector2f pos;  // the position of the terrain by pixels
    Vector2i ofs;  // the offset by size of blocks
    vector<Texture> txtr;
    vector<vector<unsigned int>> TERRAIN_DT;

public:
    Terrain(Vector2i sz, Vector2i start_pos);

    void loadTexture();
    void generateTerrain();
    inline void showTerrain();

    void setPos(Vector2f _mov);

    Vector2f getPos();  // get position
    Vector2i getOfs();  // get offset

};

Terrain::Terrain(Vector2i sz, Vector2i start_ofs) {
  TERRAIN_SZ = sz;
  ofs = start_ofs;
}

void Terrain::loadTexture() {
  string fname;
  for (int i = 0; i < sizeof(TXTR_LOAD_FNAME)/sizeof(TXTR_LOAD_FNAME[0]); ++i) {
    fname = TXTR_LOAD_FNAME[i];
    Texture t;
    if (t.loadFromFile(RES_DIR+fname+".png")) {
        txtr.push_back(t);
    } else {
        if (t.loadFromFile(RES_DIR+"no-texture"+".png")) {
            txtr.push_back(t);
        } else {
            cout << "[ERROR] Cannot load texture " << fname;
        }
    }
  }
}

void Terrain::generateTerrain() {
  for (int y = 0; y < TERRAIN_SZ.y; ++y) {
    TERRAIN_DT.push_back(vector<unsigned int>());
    for (int x = 0; x < TERRAIN_SZ.x; ++x) {
      if (x % 3 == 0 and y % 3 == 0) {
        TERRAIN_DT[y].push_back(AIR);
      } else {
        TERRAIN_DT[y].push_back(REQT);
      }
    }
  }
}


void Terrain::showTerrain() {
  RectangleShape block(Vector2f(BLOCK_SZ, BLOCK_SZ));
  cout << "CRD: " << ofs.y << " " << ofs.x << endl;
  for (int y = SCREEN_VIEW_S.y; y < SCREEN_VIEW_E.y; ++y) {
      for (int x = SCREEN_VIEW_S.x; x < SCREEN_VIEW_E.x; ++x) {
        block.setPosition(x*BLOCK_SZ+pos.x, y*BLOCK_SZ+pos.y);
        block.setTexture(&txtr[TERRAIN_DT[y-ofs.y][x-ofs.x]]);
        window.draw(block);
      }
  }
}

void Terrain::setPos(Vector2f _mov) {
  if (_mov.x == 1 and ofs.x < 0-SCREEN_VIEW_S.x) pos.x += SPEED;   // terrain movement; and to prevent the player getting out of the world
  if (_mov.y == 1 and ofs.y < 0-SCREEN_VIEW_S.y) pos.y += SPEED;
  if (_mov.x == -1 and ofs.x > -TERRAIN_SZ.x+SCREEN_VIEW_E.x) pos.x -= SPEED;
  if (_mov.y == -1 and ofs.y > -TERRAIN_SZ.y+SCREEN_VIEW_E.y) pos.y -= SPEED;

  if (pos.x/BLOCK_SZ >= 1) {  // to edit the offset (gives an illusion of movement)
      pos.x = 0;
      ofs.x += 1;
  }
  if (pos.y/BLOCK_SZ >= 1) {
      pos.y = 0;
      ofs.y += 1;
  }
  if (pos.x/BLOCK_SZ <= -1) {
      pos.x = 0;
      ofs.x -= 1;
  }
  if (pos.y/BLOCK_SZ <= -1) {
      pos.y = 0;
      ofs.y -= 1;
  }
}

Vector2f Terrain::getPos() {
  return pos;
}

Vector2i Terrain::getOfs() {
  return ofs;
}

#endif


Cookie Island/Constants.h
#ifndef HHH_CONST_H
#define HHH_CONST_H

#include <string>
#include <SFML/Graphics.hpp>

// all shape area and position is in (x, y) except for data and array(vector), it'll be flipped
// all blocks size is square

// --- GAME CONFIGURATION ---
unsigned int BLOCK_SZ = 50; // by pixels
unsigned int FPS_LIMIT = 80;
std::string LOCAL_DIR = "C:\\Users\\Andrew Shen\\Desktop\\Cookie-Island\\Cookie Island\\";
std::string RES_DIR = LOCAL_DIR+"resource\\";
sf::Vector2f SCREEN_SZ(1000, 1000);
sf::Vector2f SCREEN_VIEW_S(0, 0);  // Size in blocks; Screen View Size Start Position
sf::Vector2f SCREEN_VIEW_E(20, 20);  // Size in blocks; Screen View Size End Position
sf::RenderWindow window(sf::VideoMode(SCREEN_SZ.x, SCREEN_SZ.y), "Cookie Island");

// --- GAME MECHANICS ---
unsigned int SPEED = 5; // pixel per frame
sf::Vector2f PLAYER_SIZE(45, 90);
sf::Vector2i PLAYER_COLLISION_BOX(5, 5);  // size in block
//sf::Vector2f PLAYER_BOX_COL_OFS(0, 0);
sf::Vector2f PLAYER_ABS_POS(SCREEN_SZ.x/2, SCREEN_SZ.y/2);  // Player position to the screen



// --- GAME DATA ---
std::string TXTR_LOAD_FNAME[3] = {"air", "no-texture", "no-texture"};  // in order, followed by in block-id

#endif // CONST_H
